include/configs/xenguest_arm64.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0+
+ *
+ * (C) Copyright 2020 EPAM Systemc Inc.
+ */
+#ifndef __XENGUEST_ARM64_H
+#define __XENGUEST_ARM64_H
+
+#ifndef __ASSEMBLY__
+#include <linux/types.h>
+#endif
+#include <xen/interface/xen.h>
+
+#define CONFIG_BOARD_EARLY_INIT_F
+
+#define CONFIG_EXTRA_ENV_SETTINGS
+
+#undef CONFIG_NR_DRAM_BANKS
+/*
+ * See for reference:
+ * https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/include/public/arch-arm.h
+ */
+#define CONFIG_NR_DRAM_BANKS          GUEST_RAM_BANKS
```

Comment:
Neither **GUEST_RAM_BANKS** and **GUEST_RAM_BASE0** are part of the stable ABI. So you can't safely rely on those values in U-boot.

I am not very familiar with U-boot, would you mind to explain why does U-boot requires to know the physical address? Why can't it find out from the device-tree?

Reply:
This is because of the linker script mostly which needs to know where to put the sections

Reply:
This makes sense for ``GUEST_RAM_BASE0`` but why would the linker need ``GUEST_RAM_BANKS``?

However, I am wondering whether you can make the code PIE?

Reply:
GUEST_RAM_BANKS is needed for the u-boot itself to iterate over the predefined number of banks. Basically, you need to set CONFIG_NR_DRAM_BANKS for u-boot to be happy.
You can grep the code for the usages of that to see more
WRT PIE I am not quite sure this is an easy task for u-boot at the moment. Neither this is the target of the pull request

Reply:
See my comment https://github.com/xen-troops/u-boot/pull/1#discussion_r442093400.

You are tying U-boot to a specific version of Xen. This is potentially even a per-guest configuration depending on whether we decide to expose the host layout to the guest.

Do we really want to go down the road where you have to a new U-boot binary for each of your guest?

Reply:
Yes, but taking into account wjay u-boot expects from us I can't tell how to deal with that in some generic way with the existing code base without major changes

Reply:
The short answer is don't use anything under __XEN__. If you are planning to upstream this in U-boot, you will not make a favor to any of the user.

It would definitely be best to define a list of supported hypervisor in U-boot or try to work with Xen community to guarantee the position of the first bank in memory.

Reply:
I see another option as well:
1. Put the headers outside the u-boot tree and provide their location.
This way we can build against different versions of Xen without changing u-boot code
2. We can have a set of the relevant constants in xenguest_arm64.h surrounded with 
`#if __XEN_INTERFACE_VERSION__ < XXX
` with the version provided from outside in some way

Reply:
Or we can setup the relevant addresses vi Kconfig, but number of banks depends on live dtb, so need a predefined value anyways

Reply:
> I see another option as well:
> 
> 1. Put the headers outside the u-boot tree and provide their location.
>    This way we can build against different versions of Xen without changing u-boot code

This relies on the naming scheme to never change and that we will never provide more flexibility to the user on the layout. So it is not very ideal.

> 2. We can have a set of the relevant constants in xenguest_arm64.h surrounded with
>    `#if __XEN_INTERFACE_VERSION__ < XXX ` with the version provided from outside in some way

This would look sensible. Although, this relies on end user to hack arch-arm.h as you did in your branch. So you would want to allow the user to provide there own easily.

I will take the opportunity to note that Tianocore/EDK2 on Xen is pretty simple to use as you can compile once and then run it on every existing Xen. At the moment, all the options suggested make U-boot much less accessible to the users. This may be fine if you hack it in your side, but for a general use you need to find an easy way for the users to get things reliably working.

Reply:
> Or we can setup the relevant addresses vi Kconfig, but number of banks depends on live dtb, so need a predefined value anyways

This is a recipe for disaster. Why does U-boot need to know all the banks?

Reply:
Uboot is one of those embedded projects that typically assumes that all the information about the platform is available at *build time*. It is meant to be built tailored for a specific version of a specific board. A Uboot binary is not expected to be "portable" across different versions of the platform or different platforms. In other words, it is not expected to be portable across Xen versions.

So I don't think it is a problem to let Uboot know about Xen details at build time. Yes, that is not what we do historically but it is very much in the spirit of Uboot.

I think the question we should ask ourselves is what is the best and most flexible way to expose this information to UBoot. Xen headers? Kconfig? Device Tree (Uboot sometimes take a device tree at build time for configuration, not run time).

I don't know for sure but I suspect that using the Xen headers is better than using a Kconfig.

Reply:
Minimal set of headers + only stable ABI definitions